File: sdk/typescript/src/commands/task.ts
import { Command } from 'commander';
import { PrismerClient } from '../index';

type ClientFactory = () => PrismerClient;

export function register(parent: Command, getIMClient: ClientFactory, _getAPIClient: ClientFactory): void {
  const task = parent
    .command('task')
    .description('Manage tasks in the task marketplace');

  // task create
  task
    .command('create')
    .description('Create a new task')
    .requiredOption('--title <title>', 'task title')
    .option('--description <description>', 'task description')
    .option('--priority <priority>', 'priority: low, normal, high, urgent')
    .option('--capability <capability>', 'required agent capability')
    .option('--budget <budget>', 'budget in credits', parseFloat)
    .option('--json', 'output raw JSON response')
    .action(async (opts: {
      title: string;
      description?: string;
      priority?: string;
      capability?: string;
      budget?: number;
      json: boolean;
    }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.create({
          title: opts.title,
          description: opts.description,
          priority: opts.priority,
          requiredCapability: opts.capability,
          budget: opts.budget,
        });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`Task created successfully\n\n`);
        process.stdout.write(`ID:          ${t.id}\n`);
        process.stdout.write(`Title:       ${t.title}\n`);
        process.stdout.write(`Status:      ${t.status}\n`);
        process.stdout.write(`Priority:    ${t.priority}\n`);
        if (t.description) process.stdout.write(`Description: ${t.description}\n`);
        if (t.requiredCapability) process.stdout.write(`Capability:  ${t.requiredCapability}\n`);
        if (t.budget !== undefined) process.stdout.write(`Budget:      ${t.budget}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task list
  task
    .command('list')
    .description('List tasks')
    .option('--status <status>', 'filter by status')
    .option('--capability <capability>', 'filter by required capability')
    .option('-n, --limit <n>', 'maximum number of tasks to return', '20')
    .option('--json', 'output raw JSON response')
    .action(async (opts: {
      status?: string;
      capability?: string;
      limit: string;
      json: boolean;
    }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.list({
          status: opts.status,
          capability: opts.capability,
          limit: parseInt(opts.limit, 10),
        });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const tasks = res.data;
        if (!tasks || tasks.length === 0) {
          process.stdout.write('No tasks found.\n');
          return;
        }

        // Table header
        const idW = 24;
        const statusW = 10;
        const priorityW = 10;
        const titleW = 40;
        const header =
          'ID'.padEnd(idW) +
          'STATUS'.padEnd(statusW) +
          'PRIORITY'.padEnd(priorityW) +
          'TITLE';
        const sep = '-'.repeat(idW + statusW + priorityW + titleW);

        process.stdout.write(header + '\n');
        process.stdout.write(sep + '\n');

        for (const t of tasks) {
          const title = t.title.length > titleW ? t.title.slice(0, titleW - 3) + '...' : t.title;
          process.stdout.write(
            String(t.id).padEnd(idW) +
            String(t.status).padEnd(statusW) +
            String(t.priority).padEnd(priorityW) +
            title + '\n'
          );
        }

        process.stdout.write(`\n${tasks.length} task(s) listed.\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task get <task-id>
  task
    .command('get <task-id>')
    .description('Get task details and logs')
    .option('--json', 'output raw JSON response')
    .action(async (taskId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.get(taskId);

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`ID:           ${t.id}\n`);
        process.stdout.write(`Title:        ${t.title}\n`);
        process.stdout.write(`Status:       ${t.status}\n`);
        process.stdout.write(`Priority:     ${t.priority}\n`);
        if (t.description) process.stdout.write(`Description:  ${t.description}\n`);
        if (t.requiredCapability) process.stdout.write(`Capability:   ${t.requiredCapability}\n`);
        if (t.budget !== undefined) process.stdout.write(`Budget:       ${t.budget}\n`);
        if (t.creatorId) process.stdout.write(`Creator:      ${t.creatorId}\n`);
        if (t.assigneeId) process.stdout.write(`Assignee:     ${t.assigneeId}\n`);
        if (t.createdAt) process.stdout.write(`Created:      ${t.createdAt}\n`);
        if (t.updatedAt) process.stdout.write(`Updated:      ${t.updatedAt}\n`);
        if (t.result) process.stdout.write(`Result:       ${t.result}\n`);
        if (t.error) process.stdout.write(`Error:        ${t.error}\n`);

        const logs = t.logs ?? t.taskLogs ?? [];
        if (logs.length > 0) {
          process.stdout.write(`\nLogs (${logs.length}):\n`);
          for (const log of logs) {
            const ts = log.createdAt ?? log.timestamp ?? '';
            const msg = log.message ?? log.content ?? JSON.stringify(log);
            process.stdout.write(`  [${ts}] ${msg}\n`);
          }
        }
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task claim <task-id>
  task
    .command('claim <task-id>')
    .description('Claim a pending task')
    .option('--json', 'output raw JSON response')
    .action(async (taskId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.claim(taskId);

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`Task claimed successfully\n\n`);
        process.stdout.write(`ID:       ${t.id}\n`);
        process.stdout.write(`Title:    ${t.title}\n`);
        process.stdout.write(`Status:   ${t.status}\n`);
        process.stdout.write(`Priority: ${t.priority}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task update <task-id>
  task
    .command('update <task-id>')
    .description('Update a task')
    .option('--title <title>', 'new title')
    .option('--description <description>', 'new description')
    .option('--priority <priority>', 'new priority: low, normal, high, urgent')
    .option('--json', 'output raw JSON response')
    .action(async (taskId: string, opts: {
      title?: string;
      description?: string;
      priority?: string;
      json: boolean;
    }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.update(taskId, {
          title: opts.title,
          description: opts.description,
          priority: opts.priority,
        });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`Task updated successfully\n\n`);
        process.stdout.write(`ID:       ${t.id}\n`);
        process.stdout.write(`Title:    ${t.title}\n`);
        process.stdout.write(`Status:   ${t.status}\n`);
        process.stdout.write(`Priority: ${t.priority}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task complete <task-id>
  task
    .command('complete <task-id>')
    .description('Mark a task as complete')
    .option('--result <result>', 'result or output of the task')
    .option('--json', 'output raw JSON response')
    .action(async (taskId: string, opts: { result?: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.complete(taskId, {
          result: opts.result,
        });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`Task completed successfully\n\n`);
        process.stdout.write(`ID:     ${t.id}\n`);
        process.stdout.write(`Title:  ${t.title}\n`);
        process.stdout.write(`Status: ${t.status}\n`);
        if (t.result) process.stdout.write(`Result: ${t.result}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // task fail <task-id>
  task
    .command('fail <task-id>')
    .description('Mark a task as failed')
    .requiredOption('--error <error>', 'error message describing why the task failed')
    .option('--json', 'output raw JSON response')
    .action(async (taskId: string, opts: { error: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.tasks.fail(taskId, opts.error);

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const t = res.data;
        process.stdout.write(`Task marked as failed\n\n`);
        process.stdout.write(`ID:     ${t.id}\n`);
        process.stdout.write(`Title:  ${t.title}\n`);
        process.stdout.write(`Status: ${t.status}\n`);
        if (t.error) process.stdout.write(`Error:  ${t.error}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });
}

File: sdk/prismer-cloud/golang/cmd/prismer/task.go
package main

import (
	"fmt"

	prismer "github.com/Prismer-AI/PrismerCloud/sdk/golang"
	"github.com/spf13/cobra"
)

var taskCmd = &cobra.Command{
	Use:   "task",
	Short: "Manage tasks in the task marketplace",
}

// ── task create ─────────────────────────────────────────

var (
	taskCreateTitle       string
	taskCreateDesc        string
	taskCreateCapability  string
	taskCreateBudget      float64
	taskCreateJSON        bool
)

var taskCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a new task",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()

		opts := &prismer.CreateTaskOptions{
			Title:       taskCreateTitle,
			Description: taskCreateDesc,
			Capability:  taskCreateCapability,
			Budget:      taskCreateBudget,
		}
		res, err := client.Tasks.Create(ctx, opts)
		if err != nil {
			return err
		}
		if taskCreateJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Task created successfully\n\n")
		fmt.Printf("ID:       %v\n", data["id"])
		fmt.Printf("Title:    %v\n", data["title"])
		fmt.Printf("Status:   %v\n", data["status"])
		if taskCreateDesc != "" {
			fmt.Printf("Desc:     %v\n", data["description"])
		}
		if taskCreateCapability != "" {
			fmt.Printf("Capability: %v\n", data["capability"])
		}
		return nil
	},
}

// ── task list ───────────────────────────────────────────

var (
	taskListStatus     string
	taskListCapability string
	taskListLimit      int
	taskListJSON       bool
)

var taskListCmd = &cobra.Command{
	Use:   "list",
	Short: "List tasks",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()

		opts := &prismer.TaskListOptions{
			Capability: taskListCapability,
			Limit:      taskListLimit,
		}
		if taskListStatus != "" {
			opts.Status = prismer.TaskStatus(taskListStatus)
		}
		res, err := client.Tasks.List(ctx, opts)
		if err != nil {
			return err
		}
		if taskListJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		tasks := asList(res.Data)
		if len(tasks) == 0 {
			fmt.Println("No tasks found.")
			return nil
		}
		const idW, statusW, capW = 24, 10, 14
		header := "ID" + pad("", idW-2) + "  " + "STATUS" + pad("", statusW-6) + "  " + "CAPABILITY" + pad("", capW-10) + "  TITLE"
		fmt.Println(header)
		fmt.Println(repeatStr("-", idW+statusW+capW+30))
		for _, t := range tasks {
			tm := asMap(t)
			id := fmt.Sprintf("%v", tm["id"])
			status := fmt.Sprintf("%v", tm["status"])
			cap := fmt.Sprintf("%v", tm["capability"])
			title := fmt.Sprintf("%v", tm["title"])
			fmt.Printf("%-*s  %-*s  %-*s  %s\n", idW, id, statusW, status, capW, cap, title)
		}
		fmt.Printf("\n%d task(s) listed.\n", len(tasks))
		return nil
	},
}

// ── task get ────────────────────────────────────────────

var taskGetJSON bool
var taskGetCmd = &cobra.Command{
	Use:   "get <task-id>",
	Short: "Get task details and logs",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Tasks.Get(ctx, args[0])
		if err != nil {
			return err
		}
		if taskGetJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("ID:        %v\n", data["id"])
		fmt.Printf("Title:     %v\n", data["title"])
		fmt.Printf("Status:    %v\n", data["status"])
		if v := data["description"]; v != nil && v != "" {
			fmt.Printf("Desc:      %v\n", v)
		}
		if v := data["capability"]; v != nil && v != "" {
			fmt.Printf("Capability: %v\n", v)
		}
		if v := data["creatorId"]; v != nil && v != "" {
			fmt.Printf("Creator:   %v\n", v)
		}
		if v := data["assigneeId"]; v != nil && v != "" {
			fmt.Printf("Assignee:  %v\n", v)
		}
		if v := data["result"]; v != nil && v != "" {
			fmt.Printf("Result:    %v\n", v)
		}
		return nil
	},
}

// ── task claim ──────────────────────────────────────────

var taskClaimJSON bool
var taskClaimCmd = &cobra.Command{
	Use:   "claim <task-id>",
	Short: "Claim a pending task",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Tasks.Claim(ctx, args[0])
		if err != nil {
			return err
		}
		if taskClaimJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Task claimed successfully\n\n")
		fmt.Printf("ID:     %v\n", data["id"])
		fmt.Printf("Title:  %v\n", data["title"])
		fmt.Printf("Status: %v\n", data["status"])
		return nil
	},
}

// ── task update ─────────────────────────────────────────

var (
	taskUpdateTitle string
	taskUpdateDesc  string
	taskUpdateJSON  bool
)

var taskUpdateCmd = &cobra.Command{
	Use:   "update <task-id>",
	Short: "Update a task",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Tasks.Update(ctx, args[0], &prismer.UpdateTaskOptions{})
		if err != nil {
			return err
		}
		if taskUpdateJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Task updated\n\n")
		fmt.Printf("ID:     %v\n", data["id"])
		fmt.Printf("Title:  %v\n", data["title"])
		fmt.Printf("Status: %v\n", data["status"])
		return nil
	},
}

// ── task complete ───────────────────────────────────────

var taskCompleteResult string
var taskCompleteJSON bool
var taskCompleteCmd = &cobra.Command{
	Use:   "complete <task-id>",
	Short: "Mark a task as complete",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		opts := &prismer.CompleteTaskOptions{}
		if taskCompleteResult != "" {
			opts.Result = taskCompleteResult
		}
		res, err := client.Tasks.Complete(ctx, args[0], opts)
		if err != nil {
			return err
		}
		if taskCompleteJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Task completed\n\n")
		fmt.Printf("ID:     %v\n", data["id"])
		fmt.Printf("Title:  %v\n", data["title"])
		fmt.Printf("Status: %v\n", data["status"])
		if v := data["result"]; v != nil && v != "" {
			fmt.Printf("Result: %v\n", v)
		}
		return nil
	},
}

// ── task fail ───────────────────────────────────────────

var taskFailError string
var taskFailJSON bool
var taskFailCmd = &cobra.Command{
	Use:   "fail <task-id>",
	Short: "Mark a task as failed",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Tasks.Fail(ctx, args[0], taskFailError, nil)
		if err != nil {
			return err
		}
		if taskFailJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Task marked as failed\n\n")
		fmt.Printf("ID:     %v\n", data["id"])
		fmt.Printf("Title:  %v\n", data["title"])
		fmt.Printf("Status: %v\n", data["status"])
		return nil
	},
}

// ── helpers ─────────────────────────────────────────────

func pad(s string, n int) string {
	for len(s) < n {
		s += " "
	}
	return s
}

func repeatStr(s string, n int) string {
	result := ""
	for i := 0; i < n; i++ {
		result += s
	}
	return result
}

func init() {
	taskCreateCmd.Flags().StringVar(&taskCreateTitle, "title", "", "Task title")
	taskCreateCmd.Flags().StringVar(&taskCreateDesc, "description", "", "Task description")
	taskCreateCmd.Flags().StringVar(&taskCreateCapability, "capability", "", "Required agent capability")
	taskCreateCmd.Flags().Float64Var(&taskCreateBudget, "budget", 0, "Budget in credits")
	taskCreateCmd.Flags().BoolVar(&taskCreateJSON, "json", false, "Output raw JSON")
	_ = taskCreateCmd.MarkFlagRequired("title")

	taskListCmd.Flags().StringVar(&taskListStatus, "status", "", "Filter by status")
	taskListCmd.Flags().StringVar(&taskListCapability, "capability", "", "Filter by required capability")
	taskListCmd.Flags().IntVarP(&taskListLimit, "limit", "n", 20, "Maximum number of tasks to return")
	taskListCmd.Flags().BoolVar(&taskListJSON, "json", false, "Output raw JSON")

	taskGetCmd.Flags().BoolVar(&taskGetJSON, "json", false, "Output raw JSON")

	taskClaimCmd.Flags().BoolVar(&taskClaimJSON, "json", false, "Output raw JSON")

	taskUpdateCmd.Flags().StringVar(&taskUpdateTitle, "title", "", "New title")
	taskUpdateCmd.Flags().StringVar(&taskUpdateDesc, "description", "", "New description")
	taskUpdateCmd.Flags().BoolVar(&taskUpdateJSON, "json", false, "Output raw JSON")

	taskCompleteCmd.Flags().StringVar(&taskCompleteResult, "result", "", "Result or output of the task")
	taskCompleteCmd.Flags().BoolVar(&taskCompleteJSON, "json", false, "Output raw JSON")

	taskFailCmd.Flags().StringVar(&taskFailError, "error", "", "Error message describing why the task failed")
	taskFailCmd.Flags().BoolVar(&taskFailJSON, "json", false, "Output raw JSON")
	_ = taskFailCmd.MarkFlagRequired("error")

	taskCmd.AddCommand(taskCreateCmd)
	taskCmd.AddCommand(taskListCmd)
	taskCmd.AddCommand(taskGetCmd)
	taskCmd.AddCommand(taskClaimCmd)
	taskCmd.AddCommand(taskUpdateCmd)
	taskCmd.AddCommand(taskCompleteCmd)
	taskCmd.AddCommand(taskFailCmd)
	rootCmd.AddCommand(taskCmd)
}
